go/chat/storage/storage.go

```diff
@@ -425,38 +424,97 @@ func (s *Storage) updateAllSupersededBy(ctx context.Context, convID chat1.Conver
+		getMsg := func(msgID chat1.MessageID) ([]chat1.MessageUnboxed, Error) {
 			rc := NewSimpleResultCollector(1)
-			err := s.engine.ReadMessages(ctx, rc, convID, uid, supersededID)
+			err := s.engine.ReadMessages(ctx, rc, convID, uid, msgID)
 			if err != nil {
 				// If we don't have the message, just keep going
 				if _, ok := err.(MissError); ok {
-					continue
+					return nil, nil
+				}
+				return nil, err
+			}
+			return rc.Result(), nil
+		}
+		// helper to update a reaction's target when the reaction itself is deleted
+		updateReactionTarget := func(reactionMsg chat1.MessageUnboxed) (*chat1.MessageUnboxed, Error) {
+			s.Debug(ctx, "updateReactionTarget: reationMsg:%v", reactionMsg)
+			if reactionMsg.Valid().MessageBody.IsNil() {
+				return nil, nil
+			}
+
+			targetMsgID := reactionMsg.Valid().MessageBody.Reaction().MessageID
+			targetMsgs, err := getMsg(targetMsgID)
+			if err != nil || targetMsgs == nil {
+				return nil, err
+			}
+			targetMsg := targetMsgs[0]
+			if targetMsg.IsValid() {
+				mvalid := targetMsg.Valid()
+				reactionIDs := []chat1.MessageID{}
+				for _, msgID := range mvalid.ServerHeader.ReactionIDs {
+					if msgID != reactionMsg.GetMessageID() {
+						reactionIDs = append(reactionIDs, msgID)
+					}
 				}
+				mvalid.ServerHeader.ReactionIDs = reactionIDs
+				newMsg := chat1.NewMessageUnboxedWithValid(mvalid)
+				return &newMsg, nil
+			}
+			return nil, nil
+		}
+
+		// Set all supersedes targets
+		for _, supersededID := range supersededIDs {
+			s.Debug(ctx, "updateSupersededBy: supersedes: id: %d supersedes: %d", msgid, supersededID)
+			// Read superseded msg
+			superMsgs, err := getMsg(supersededID)
+			if err != nil {
 				return err
 			}
-			superMsgs := rc.Result()
 			if len(superMsgs) == 0 {
 				continue
 			}
 
-			// Update supersededBy on the target message if we have it. And if
-			// the superseder is a deletion, delete the body as well.
+			// Update supersededBy and reactionIDs on the target message if we
+			// have it. If the superseder is a deletion, delete the body as
+			// well. If we are deleting a reaction, update the reaction's
+			// target message.
 			superMsg := superMsgs[0]
 			if superMsg.IsValid() {
 				s.Debug(ctx, "updateSupersededBy: writing: id: %d superseded: %d", msgid, supersededID)
 				mvalid := superMsg.Valid()
-				mvalid.ServerHeader.SupersededBy = msgid
+
+				// reactions don't update SupersededBy, instead they rely on ReactionIDs
+				if msg.GetMessageType() == chat1.MessageType_REACTION {
+					mvalid.ServerHeader.ReactionIDs = s.updateReactionIDs(mvalid.ServerHeader.ReactionIDs, msgid)
+				} else {
+					mvalid.ServerHeader.SupersededBy = msgid
+				}
+
+				var newMsg chat1.MessageUnboxed
 				if msg.GetMessageType() == chat1.MessageType_DELETE {
+					// We have to find the message we are reacting to and
+					// update it's ReactionIDs as well.
+					if superMsg.GetMessageType() == chat1.MessageType_REACTION {
```

Comment:
Like we talked about in person, I think we can just remove this and rely on the logic in supersedes.go